src/dynamic_set.rs

```diff
@@ -38,71 +37,56 @@ fn entry_alignment_is_sufficient() {
-        Set {
-            buckets: unsafe { Box::from_raw(vec.as_ptr() as *mut [T; NB_BUCKETS]) },
-        }
-    })
+pub(crate) static DYNAMIC_SET: Lazy<Set> = Lazy::new(|| {
+    // NOTE: Using const initialization for buckets breaks the small-stack test.
+    // ```
+    // // buckets: [Mutex<Vec<Box<Entry>>>; NB_BUCKETS],
+    // const MUTEX_VEC: Mutex<Vec<Box<Entry>>> = Mutex::new(vec![]);
+    // let buckets = Box::new([MUTEX_VEC; NB_BUCKETS]);
+    // ```
+    let buckets = (0..NB_BUCKETS).map(|_| Mutex::new(vec![])).collect();
+    Set { buckets }
 });
 
 impl Set {
-    pub(crate) fn insert(&mut self, string: Cow<str>, hash: u32) -> NonNull<Entry> {
+    pub(crate) fn insert(&self, string: Cow<str>, hash: u32) -> NonNull<Entry> {
         let bucket_index = (hash & BUCKET_MASK) as usize;
+        let mut vec = self.buckets[bucket_index].lock();
+        if let Some(entry) = vec
+            .iter_mut()
+            .find(|e| e.hash == hash && *e.string == *string)
         {
-            let mut ptr: Option<&mut Box<Entry>> = self.buckets[bucket_index].as_mut();
-
-            while let Some(entry) = ptr.take() {
-                if entry.hash == hash && *entry.string == *string {
-                    if entry.ref_count.fetch_add(1, SeqCst) > 0 {
-                        return NonNull::from(&mut **entry);
-                    }
-                    // Uh-oh. The pointer's reference count was zero, which means someone may try
-                    // to free it. (Naive attempts to defend against this, for example having the
-                    // destructor check to see whether the reference count is indeed zero, don't
-                    // work due to ABA.) Thus we need to temporarily add a duplicate string to the
-                    // list.
-                    entry.ref_count.fetch_sub(1, SeqCst);
-                    break;
-                }
-                ptr = entry.next_in_bucket.as_mut();
+            if entry.ref_count.fetch_add(1, SeqCst) > 0 {
+                return NonNull::from(&mut **entry);
             }
+            // Uh-oh. The pointer's reference count was zero, which means someone may try
+            // to free it. (Naive attempts to defend against this, for example having the
+            // destructor check to see whether the reference count is indeed zero, don't
+            // work due to ABA.) Thus we need to temporarily add a duplicate string to the
+            // list.
+            entry.ref_count.fetch_sub(1, SeqCst);
         }
         debug_assert!(mem::align_of::<Entry>() >= ENTRY_ALIGNMENT);
         let string = string.into_owned();
         let mut entry = Box::new(Entry {
-            next_in_bucket: self.buckets[bucket_index].take(),
             hash,
             ref_count: AtomicIsize::new(1),
             string: string.into_boxed_str(),
         });
         let ptr = NonNull::from(&mut *entry);
-        self.buckets[bucket_index] = Some(entry);
-
+        vec.push(entry);
```

Comment:
The existing code adds the entry to the head of the linked list.
This opens the possibility for the next `insert()` usage to again `find()` the entry that is about to be removed, and we'll end up inserting another duplicate entry (i.e. an unintended 3rd, 4th etc.).
With the existing implementation, the next `insert()` is guaranteed to `find()` the new entry that was added (and increment its ref count) since it's at the front.